EUIfy Console - partially de-angularize and move custom views out of top_nav

lizozom opened the pull request:
## Release note

Console's welcome panel, help panel, and settings form UIs have all been updated with a new look that is more consistent with the rest of Kibana's UI. Changes to the layout of these elements also improves the user experience of using Console on smaller screens.

## Summary

This PR is part of the task of [moving kbn_top_nav to new platform](https://github.com/elastic/kibana/issues/31968#issuecomment-503053729).
To do so, I wanted to narrow the functionality of `kbn_top_nav`, and specifically, stop using it for rendering inline custom views (examples below).

This PR focuses on the Sense (DevTools) app, which was the heaviest user of this functionality. I will create additional PRs for graph and timelion, that make smaller use of this ability.

I was also able to delete `SenseTopNavController` which was the only remaining custom `TopNavController` implementation

These are the UI changes in this PR

### Welcome notification
Welcome notification used to be shown _automatically_ (i.e. without the user clicking on any navigation button) as part of the navigation menu.
It now uses a **flyout**.
If it's closed by clicking on the X or anywhere on the screen, it will come back next time the user refreshes the window. However, if the user clicks on the 'Dismiss' button, it won't be shown again.
The only changes here are **the display location** and **the text on the button**, which I felt was more meaningful.

![image](https://user-images.githubusercontent.com/3016806/59851344-1720e800-9375-11e9-9ab6-18a262e8d968.png)

### Help

Help used to be displayed inside a **flyout**.
There were no other changes to the UI.

![image](https://user-images.githubusercontent.com/3016806/59851556-9ca49800-9375-11e9-929f-e8529aecc5b2.png)

### Settings

Previously, DevTools settings were displayed inline, inside the navigation menu.
I moved them to a **modal**. 

#### Important
Coincidentally, @cjcenizal merged two PRs yesterday, and I had to re-implement them partially. They are tested and seem to work correctly.
- [Allow users to disable polling in Console](https://github.com/elastic/kibana/pull/38949)
- [Add setting for disabling use of triple quotes in the Console output pane](https://github.com/elastic/kibana/pull/39114)

![image](https://user-images.githubusercontent.com/3016806/59851742-1472c280-9376-11e9-89a5-2950194ca97f.png)

### History

History console still **opens inline**, but not as part of the navigation menu, as I felt like moving it to a modal \ flyout worsens the UX.
I was able to achieve a small usability improvement, where if history is open, running queries **are added immediately to the list** (before you had to re-open the history to see changes).

![image](https://user-images.githubusercontent.com/3016806/59851922-6ddaf180-9376-11e9-8cab-754a0c6712a2.png)

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

lizozom commented:
@AlonaNadler @elastic/eui-design I would appreciate a review of the UI changes made here.

Please note that my goal was **not** to create the final perfect UI (as this app still needs to be Reactified and EUIficated), but to clear out top nav quickly, while getting some quick EUI wins.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2990/)

elasticmachine commented:
Pinging @elastic/es-ui

cchaos commented on src/legacy/core_plugins/console/public/src/components/help_panel.tsx:
```diff
@@ -0,0 +1,152 @@
+
+export function OpenHelpPanel(props: Props) {
+  return (
+    <EuiFlyout ownFocus onClose={props.onClose} data-test-subj="loadSearchForm">
```
Add the prop `maxWidth` on here because the line-length for the text gets way too long for readability.

Same for the welcome flyout.

lizozom replied:
Set to 700px

cchaos commented on src/legacy/core_plugins/console/public/src/components/help_panel.tsx:
```diff
@@ -0,0 +1,152 @@
+        </EuiTitle>
+      </EuiFlyoutHeader>
+      <EuiFlyoutBody>
+        <EuiPanel>
```
Remove panel. This already resides in container (flyout) it doesn't need the border.

cchaos commented on src/legacy/core_plugins/console/public/src/components/help_panel.tsx:
```diff
@@ -0,0 +1,152 @@
+      </EuiFlyoutHeader>
+      <EuiFlyoutBody>
+        <EuiPanel>
+          <EuiFlexGroup>
```
Remove this wrapping FlexGroup

cchaos commented on src/legacy/core_plugins/console/public/src/components/help_panel.tsx:
```diff
@@ -0,0 +1,152 @@
+      <EuiFlyoutBody>
+        <EuiPanel>
+          <EuiFlexGroup>
+            <EuiFlexItem className="euiText euiText--medium euiText--constrainedWidth">
```
This is element should just be `EuiText` which will add all those classes.

cchaos commented on src/legacy/core_plugins/console/public/src/components/help_panel.tsx:
```diff
@@ -0,0 +1,152 @@
+                />
+              </h3>
+              <EuiSpacer />
+              <span>
```
```suggestion
              <p>
```

cchaos commented on src/legacy/core_plugins/console/public/src/components/settings_modal.tsx:
```diff
@@ -0,0 +1,263 @@
+        <EuiModalBody className="euiText euiText--medium euiText--constrainedWidth">
+          <EuiFlexGrid columns={1}>
+            <EuiFlexItem>
+              <EuiFlexItem>
```
All of these EuiFlexItem and EuiFlexGrid wrappers are unnecessary since you're just doing a straight column which is the normal box model.

cchaos commented on src/legacy/core_plugins/console/public/src/components/settings_modal.tsx:
```diff
@@ -0,0 +1,263 @@
+                </EuiText>
+              </EuiFlexItem>
+              <EuiFlexItem>
+                <EuiFieldNumber
```
This should be wrapped in a `<EuiFormRow label="Font size">`.  See http://localhost:8030/#/forms/form-layouts for more info on that component.

cchaos commented on src/legacy/core_plugins/console/public/src/components/settings_modal.tsx:
```diff
@@ -0,0 +1,263 @@
+                  id="console.settingsPage.jsonSyntaxLabel"
+                  defaultMessage="JSON syntax"
+                />
+              </h4>
```
Wrap all these h4's in a `<EuiTitle size="xs">` for styling

lizozom replied:
Used formrow labels instead

cchaos commented on src/legacy/core_plugins/console/public/src/components/settings_modal.tsx:
```diff
@@ -0,0 +1,263 @@
+                  defaultMessage="Autocomplete"
+                />
+              </h4>
+              <EuiSwitch
```
These should actually be changed back to checkboxes and wrapped in a legend style form row like:

```jsx
<EuiFormRow
  label="Autocomplete"
  labelType="legend">
          <EuiCheckboxGroup
            options={this.state.checkboxes}
            idToSelectedMap={this.state.checkboxIdToSelectedMap}
            onChange={this.onCheckboxChange}
          />
</EuiFormRow>
```

cchaos commented on src/legacy/core_plugins/console/public/src/components/settings_modal.tsx:
```diff
@@ -0,0 +1,263 @@
+      >
+        <EuiModalHeader>
+          <EuiModalHeaderTitle>
+            <FormattedMessage id="console.settingsPage.pageTitle" defaultMessage="Settings" />
```
```suggestion
            <FormattedMessage id="console.settingsPage.pageTitle" defaultMessage="Console settings" />
```

cchaos commented on src/legacy/core_plugins/console/public/src/components/settings_modal.tsx:
```diff
@@ -0,0 +1,263 @@
+          </EuiModalHeaderTitle>
+        </EuiModalHeader>
+
+        <EuiModalBody className="euiText euiText--medium euiText--constrainedWidth">
```
Remove these classes and only wrap the necessary paragraphs of text with `EuiText`

cchaos commented on src/legacy/core_plugins/console/public/src/components/settings_modal.tsx:
```diff
@@ -0,0 +1,263 @@
+                id="wrapLines"
+                label={
+                  <FormattedMessage
+                    defaultMessage="Wrap Long Lines"
```
Watch your casing, we always use sentence case

```suggestion
                    defaultMessage="Wrap long lines"
```

cchaos commented on src/legacy/core_plugins/console/public/src/components/settings_modal.tsx:
```diff
@@ -0,0 +1,263 @@
+                  defaultMessage="Refreshing autocomplete suggestions"
+                />
+              </h4>
+              <FormattedMessage
```
Wrap this in a `<EuiText><p>`

lizozom replied:
Still needed?

cjcenizal replied:
I think this is fine as-is?

cchaos commented on src/legacy/core_plugins/console/public/src/components/welcome_panel.tsx:
```diff
@@ -0,0 +1,146 @@
+        </EuiTitle>
+      </EuiFlyoutHeader>
+      <EuiFlyoutBody>
+        <EuiPanel>
```
Remove panel

cchaos commented on src/legacy/core_plugins/console/public/src/components/welcome_panel.tsx:
```diff
@@ -0,0 +1,146 @@
+      <EuiFlyoutBody>
+        <EuiPanel>
+          <EuiFlexGroup className="euiText" direction="column">
+            <EuiFlexItem>
```
Remove all EuiFlexItem and EuiFlexGroup wrappers. Instead, wrap in `EuiText`

cchaos commented on src/legacy/core_plugins/console/public/src/components/welcome_panel.tsx:
```diff
@@ -0,0 +1,146 @@
+                  defaultMessage="Quick intro to the UI"
+                />
+              </h4>
+              <div>
```
```suggestion
              <p>
```

cchaos commented on src/legacy/core_plugins/console/public/src/components/welcome_panel.tsx:
```diff
@@ -0,0 +1,146 @@
+                the response pane on the right side."
+                />
+              </div>
+              <div>
```
```suggestion
              <p>
```

cchaos commented on src/legacy/core_plugins/console/public/src/components/welcome_panel.tsx:
```diff
@@ -0,0 +1,146 @@
+                />
+              </div>
+              <EditorExample />
+              <div>
```
```suggestion
              <p>
```

cchaos commented on src/legacy/core_plugins/console/public/src/components/welcome_panel.tsx:
```diff
@@ -0,0 +1,146 @@
+              </ul>
+
+              <div>
+                <EuiButton
```
Move this button to a `EuiFlyoutFooter`

cchaos commented on src/legacy/core_plugins/console/public/src/directives/_history.scss:
```diff
@@ -1,3 +1,7 @@
+sense-history {
+  padding: 8px;
```
When adding SASS, please use EUI sass variables instead of hard-coded px values. We have a [guides page on SASS](http://localhost:8030/#/guidelines/sass) to help you out.

For this one it would be:
```suggestion
  padding: $euiSizeS;
```

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3018/)

cjcenizal commented on src/legacy/core_plugins/console/public/src/input.js:
```diff
@@ -139,6 +139,7 @@ export function initializeInput($el, $actionsEl, $copyAsCurlEl, output, openDocu
 
             // we have someone on the other side. Add to history
             history.addToHistory(esPath, esMethod, esData);
+            addedToHistoryCb();
```
On around line 215, a keyboard shortcut is added that lets the user hit Ctrl-Enter to execute whichever request their cursor is inside of. This means `sendCurrentRequestToES` is called when the user hits Ctrl-Enter. However, the keyboard-shortcut code passes in a non-function argument as the `addedToHistoryCb ` parameter, and this causes a silent failure when line 142 attempts to call it. This silent failure causes no results to show up in the response pane, despite the request successfully being executed.

We can fix this by first wrapping this in a condition:
```js
if (addedToHistoryCb) {
  addedToHistoryCb();
}
```

And then on line 215 we can change the keyboard shortcut to call`sendCurrentRequestToES` without arguments:

```js
  input.commands.addCommand({
    name: 'send to elasticsearch',
    bindKey: { win: 'Ctrl-Enter', mac: 'Command-Enter' },
    // Don't pass any arguments to sendCurrentRequestToES.
    exec: () => sendCurrentRequestToES()
  });
```

cjcenizal commented on src/legacy/core_plugins/console/public/src/components/help_show_panel.tsx:
```diff
@@ -17,26 +17,31 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import { I18nContext } from 'ui/i18n';
+import { OpenHelpPanel } from './help_panel';
```
Nit: Can we remove the verb from this name by changing `OpenHelpPanel` to `HelpPanel`? I'm used to seeing verbs used in method/function names, and nouns in classes and class instances. :)

lizozom replied:
No idea why I did this!
Think I copied it from somewhere O_O

cjcenizal commented on src/legacy/core_plugins/console/public/src/components/help_show_panel.tsx:
```diff
@@ -17,26 +17,31 @@
+let isOpen = false;
 
-const storage = require('../storage');
+export function helpShowPanel() {
```
By the same token, I'd find `showHelpPanel` easier to grok in this case because it's a function.

cjcenizal commented on src/legacy/core_plugins/console/public/src/controllers/sense_controller.js:
```diff
@@ -94,7 +114,9 @@ module.controller('SenseController', function SenseController(Private, $scope, $
     input.focus();
-    input.sendCurrentRequestToES();
+    input.sendCurrentRequestToES(() => {
+      $scope.historyDirty = new Date().getTime();
```
Could you walk me through the role `$scope.historyDirty` plays in the UX?

lizozom replied:
It's used to force the `sense-history` to regenerate in case a user runs a query, while the history panel is open.

cjcenizal replied:
Ah, I see, thanks! Could we please rename it to `lastRequestTimestamp`, and add a comment on this line:

```js
// History watches this value and will re-render itself when it changes, so that
// the list of requests stays up-to-date as new requests are sent.
```

cjcenizal commented:
I just did a QA pass and verified:

- [ ] Keyboard shortcuts work (except for executing the selected request)
- [ ] Gear menu options work
- [ ] Executing multiple requests in sequence works
- [ ] History panel works
- [ ] Settings panel works including:
  - [ ] Polling settings
  - [ ] Manual refresh button
  - [ ] Disabling triple quotes

lizozom commented:
@cjcenizal I removed the overlay from the flyout.

However, I actually don't think that adding a spacer there will help anything, just take even more space off the screen. I loved the new design @cchaos suggested for the history panel, but I don't have enough time on this task to implement it. Could we just leave it as is?

Fixed all other comments!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3414/)

lizozom commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3415/)

cjcenizal commented on src/legacy/core_plugins/console/public/src/components/settings_modal.tsx:
```diff
@@ -0,0 +1,272 @@
+            />
+          </EuiFormRow>
+
+          <EuiFlexItem grow={false}>
```
Can we remove this wrapping `EuiFlexItem` element? It causes the button to go full-width, which I think over-emphasizes it. This looks better to me:

![image](https://user-images.githubusercontent.com/1238659/60059484-9d08a000-96a1-11e9-8493-e28f02cb9529.png)

cjcenizal commented on src/legacy/core_plugins/console/public/src/components/help_panel.tsx:
```diff
@@ -0,0 +1,144 @@
+              defaultMessage="Keyboard commands"
+            />
+          </h3>
+          <EuiSpacer />
```
This spacer isn't needed.

cjcenizal commented on src/legacy/core_plugins/console/public/src/controllers/sense_controller.js:
```diff
@@ -35,10 +40,16 @@ module.run(function ($rootScope) {
 });
 
+function showWelcomeMessageIfNeeded() {
+  if (true || storage.get('version_welcome_shown') !== '@@SENSE_REVISION') {
```
I think `true` here is leftover from testing? Because now the welcome flyout is always shown when I navigate to Console. 😄

lizozom replied:
oops!

cjcenizal commented on src/legacy/core_plugins/console/public/src/components/settings_modal.tsx:
```diff
@@ -0,0 +1,272 @@
+    <EuiOverlayMask>
+      <EuiModal
+        data-test-subj="devToolsSettingsModal"
+        className="devToolsSettingsModal"
```
Is this classname defined anywhere? I can't seem to find it.

lizozom commented:
@cjcenizal added logic to close side panels on `$scope.$on('$destroy',...)`

lizozom commented:
@cjcenizal done

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3765/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3796/)

cjcenizal commented:
Retest

silne30 commented:
@lizozom I went ahead and ran through these changes and did not notice any regressions. I tested this on Firefox and Chrome and also tested the changes made due to feedback from others. Nice work! Getting a VM up and running to run through it on IE11.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3906/)

lizozom commented on src/legacy/core_plugins/console/public/src/helpers/welcome_show_panel.tsx:
```diff
@@ -26,16 +26,11 @@ import storage from '../storage';
-  const container = document.createElement('div');
+export function welcomeShowPanel(): () => void {
   const onClose = () => {
+    if (!container) return;
```
TypeScript :-(
getElementById may return undefined...

cjcenizal replied:
Thanks, TypeScript. :P

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3951/)

lizozom commented:
@cchaos could you review the final design?

lizozom commented:
> @lizozom I went ahead and ran through these changes and did not notice any regressions. I tested this on Firefox and Chrome and also tested the changes made due to feedback from others. Nice work! Getting a VM up and running to run through it on IE11.

Waiting for your feedback!

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4092/)
